src/Forms/RestoreAction.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace SilverStripe\Forms;
+
+use SilverStripe\ORM\ValidationException;
+use SilverStripe\Versioned\Versioned;
+
+/**
+ * Generic restore action to restore an archived item to draft
+ */
+class RestoreAction
+{
+    /**
+     * Returns a message which notifies the user of a successful restoration
+     * and if anything has changed
+     *
+     * @param Object $item
+     * @return Array $message
+     */
+    public static function restore($item)
+    {
+        $isOnDraft = $item->isOnDraft();
+        $isPublished = $item->isPublished();
+        $canEdit = $item->canEdit();
+
+        if (!$canEdit) {
+            throw new ValidationException(
+                _t(
+                    'SilverStripe\\Admin\\ArchiveAdmin.RESTORE_FALIURE_PERMISSION',
+                    'Insufficient permission to restore item'
+                )
+            );
+        }
+
+        if ($isOnDraft || $isPublished) {
+            throw new ValidationException(
+                _t(
+                    'SilverStripe\\Admin\\ArchiveAdmin.RESTORE_FALIURE_STATE',
+                    'This item already exists and cannot be restored'
+                )
+            );
+        }
+
+        $classname = $item->classname;
+        $id = $item->ID;
+
+        if (!$classname || !$id) {
+            return new ValidationException("Unable to restore item", 400);
+        }
+
+        $changedLocation = self::shouldRestoreToRoot($item);
+
+        /** @var DataObject $restoredItem */
+        $archivedItem = Versioned::get_latest_version($classname, $id);
+
+        if (!$archivedItem) {
+            return new ValidationException($classname . " #$id not found", 400);
+        }
+
+        if (method_exists($archivedItem, 'doRestoreToStage')) {
+            $restoredItem = $archivedItem->doRestoreToStage();
+        } else {
+            $archivedItem->writeToStage(Versioned::DRAFT);
```

Comment:
I think you can get away with just calling `write()`. That's [what I told people to do in the version doc](https://docs.silverstripe.org/en/4/developer_guides/model/versioning/#restoring-an-archived-version).